src/rt/sched/behaviourcore.h

```diff
@@ -735,178 +740,226 @@ namespace verona::rt
-
-          if (cown->read_ref_count.try_write())
-          {
-            Logging::cout() << " Writer at head of queue and got the cown "
-                            << *curr_slot << Logging::endl;
-            ec[std::get<0>(indexes[first_chain_index])]++;
-            yield();
-            continue;
-          }
-
-          Logging::cout() << " Writer waiting for previous readers cown "
-                          << *curr_slot << Logging::endl;
-          yield();
-          cown->next_writer = body;
+          std::get<4>(chain_info[i]) = true;
           continue;
         }
 
-        // Release any transferred count
-        acquire_with_transfer(cown, transfer_count, 0);
-
-        yield();
-        Logging::cout() << " Someone in queue cown " << *curr_slot
-                        << " previous " << *prev_slot << Logging::endl;
-
         while (prev_slot->is_wait_2pl())
         {
           Systematic::yield_until(
             [prev_slot]() { return !prev_slot->is_wait_2pl(); });
           Aal::pause();
         }
 
-        if (curr_slot->is_read_only())
+        if (new_slot->is_read_only())
         {
-          if (prev_slot->set_next_slot_reader(curr_slot))
+          if (prev_slot->set_next_slot_reader(new_slot))
           {
             Logging::cout()
               << " Previous slot is a writer or blocked reader cown "
-              << *curr_slot << Logging::endl;
+              << *new_slot << Logging::endl;
             yield();
             continue;
           }
 
           yield();
           bool first_reader = cown->read_ref_count.add_read();
-          Logging::cout() << " Reader got the cown " << *curr_slot
+          Logging::cout() << " Reader got the cown " << *new_slot
                           << Logging::endl;
           yield();
-          curr_slot->set_read_available();
-          ec[std::get<0>(indexes[first_chain_index])]++;
+          new_slot->set_read_available();
+          ec[first_body_index]++;
           if (first_reader)
           {
             Logging::cout()
               << "Acquiring reference count for first reader on cown "
-              << *curr_slot << Logging::endl;
+              << *new_slot << Logging::endl;
             Cown::acquire(cown);
           }
           continue;
         }
 
         Logging::cout()
           << " Writer waiting for cown. Set next of previous slot cown "
-          << *curr_slot << " previous " << *prev_slot << Logging::endl;
+          << *new_slot << " previous " << *prev_slot << Logging::endl;
         prev_slot->set_next_slot_writer(first_body);
         yield();
       }
 
-      // Second phase - Release phase.
+      // Third phase - Release phase.
       for (size_t i = 0; i < body_count; i++)
       {
         Logging::cout() << "Release phase for behaviour " << bodies[i]
                         << Logging::endl;
       }
-      for (size_t i = 0; i < count; i++)
+      // FIXME: Shouldn't this be only over chains?
```

Comment:
Agreed, we should only be marking the last slot of each chain as ready.  The 
```
if (slot->is_wait_2pl())
```
means this will only apply to the final elements in a chain, but lets use the better data we now have in the prepare phase.